Fix the_content link replacements

alexwpengine opened the pull request:
## Tasks

- [x] I have signed a [Contributor License Agreement (CLA)](https://github.com/wpengine/faustjs#contributor-license-agreement) with WP Engine.
- [x] If a code change, I have written testing instructions that the whole team & outside contributors can understand.
- [x] I have written and included a comprehensive [changeset](https://github.com/wpengine/faustjs/blob/canary/DEVELOPMENT.md#deployment) to properly document the changes I've made.

## Description
Fix duplicate links and port not taken into account.
 * Replace the URLs in the post content basesd on two settings if enabled:
 1. Enable Post and Category URL rewrites - rewrite the WordPress internal URL in the content with the Front-end URL
1.1 If NOT enabled - use the WordPress URL
2. Use the WordPress domain for media URLs in post content - use WordPress URL for media files (images,csv,pdf)
2.1 If NOT enabled - use front-end url
## Related Issue(s):

Provide the GitHub issue(s) number for issue tracking purposes, use the following syntax:

- #2096 
https://github.com/wpengine/faustjs/pull/2014

## Testing


Sample post with images, csv/pdf, an internal link. 

The expected result must be:
 1. Enable Post and Category URL rewrites - rewrite the WordPress internal URL in the content with the Front-end URL
1.1 If NOT enabled - use the WordPress URL
2. Use the WordPress domain for media URLs in post content - use WordPress URL for media files (images,csv,pdf)
2.1 If NOT enabled - use front-end url


## Screenshots

<!--
If this is a visual change include relevant screenshots about the behavior of the application before and after this change.
-->

## Documentation Changes

<!--
List corresponding changes to the documentation.
-->

## Dependant PRs

<!--
List any dependent PR's that are awaiting review. Use the following syntax:

- #1234
-->

alexwpengine committed:
Fix the_content link replacements

alexwpengine committed:
simplify the code

alexwpengine committed:
fix formatting

alexwpengine committed:
add fullstops

alexwpengine committed:
fix host_url port

alexwpengine committed:
fix empty space

alexwpengine committed:
fix slash issue

colinmurphy commented:
@alexwpengine 

I will try and have a look tomorrow at why those tests are failing.

colinmurphy committed:
Merge branch 'heads/canary' into 2096-faustwp---content-replacement-bug

colinmurphy committed:
Updated e2e tests. Delete directories only if they exist for e2e tests.

colinmurphy commented:
@alexwpengine @theodesp 

I believe the e2e issue is related to - https://github.com/wpengine/faustjs/pull/2089

colinmurphy committed:
Updated e2e tests for example tutorial to use npm degit

As per issue #2089 there is a cache issue with using npx create app for the example tutorial on Github's side. Not sure why it is only failing for this branch though.

colinmurphy committed:
Fixed npm degit issue

colinmurphy committed:
Fixing npm degit.

colinmurphy committed:
Install e2e dependencies

colinmurphy committed:
Fixing npm ci issue for e2e directory

colinmurphy committed:
Updated e2e example test to only run when there is a change within the example directory

colinmurphy commented on plugins/faustwp/includes/replacement/callbacks.php:
```diff
@@ -50,16 +55,42 @@ function content_replacement( ?string $content ) {
+			$relative = str_replace( reset( $is_wp_url ), '', $url );
+			$updated  = 'href="' . $frontend_uri . $relative . '"';
+
+			$original = $href_links[0][ $i ];
```
@alexwpengine 

I think we should check that key exists before doing any string replace

colinmurphy reviewed (changes requested):
@alexwpengine 

I fixed the e2e bug.

Can we also add tests for what this fixes too please. Thank you for fixing this but I think it would be useful to have new tests too which act as documentation too.

alexwpengine committed:
minor improvements

alexwpengine committed:
fix linting

alexwpengine committed:
adding port tests

colinmurphy reviewed (approved):
@alexwpengine 

This PR looks great 🚀 🚀 🚀 

Thanks for adding in the tests and also for fixing the issues with ports ❤️

colinmurphy commented:
@alexwpengine 

Can you also add a changeset for this too please.  You should be able to add one clicking on the link above but let me know if you need any help.

alexwpengine committed:
Create long-laws-doubt.md
